### ISSUE TITLE
JsonResponse setContent, setEncodingOptions

### ISSUE THREAD

**Issue by kanariezwart** [symfony/symfony#16541]
Consider when creating a `JsonResponse` and calling `setContent()` instead of `setData()` (bypassing `this->data` and meant for adding serialized content straight  onto `$this->content` and skipping the added json_encode feature. 

``` php
$jsonData = '{...}'; // something something
$response = new JsonResponse();
$response->setContent($jsonData)
$response->setEncodingOptions(JSON_PRETTY_PRINT); // some encoding option

return $response;
```

This will return an empty string as the JsonResponse content. Of course when setting content manually, there is no need for setting options anymore. However it's not always clear when data or content is passed to a JsonResponse. Inside the `setEncodingOptions()`:

``` php
    // Symfony\Component\HttpFoundation/JsonReponse.php
    public function setEncodingOptions($encodingOptions)
    {
        $this->encodingOptions = (int) $encodingOptions;

        return $this->setData(json_decode($this->data));
    } 
```

This also triggers the re-adding of `this->data` as content. I think this leads to enexpected behaviour.  This proposal might not be the prettiest, but it works:

``` php
    // Symfony\Component\HttpFoundation/JsonReponse.php
    public function setEncodingOptions($encodingOptions)
    {
        $this->encodingOptions = (int)$encodingOptions;
        if ($this->content !== null) {
           $this->data = $this->content;
        }
        return $this->setData(json_decode($this->data));
    }
```

Another option to get rid of this issue is when calling `setContent()`,  try to set and isFinal flag or something.

``` php
// Symfony\Component\HttpFoundation/JsonReponse.php
public function setContent($content)
{
  $this->data =  json_decode($content);
 // call Symfony\Component\HttpFoundation/Reponse.php
  parent::setContent($content);
}
```

**Comment by kanariezwart** [comment]
Any thoughts?

**Comment by ro0NL** [comment]
I currently use the generic `Response` object for ready-to-use JSON strings, as I dont like the overhead of json_decode/encode in the `JsonResponse` object, ie; `setData(json_decode($contents))`, for something I already have.

Also combining setContent and setCallback leads to unexpected behaviour.. as the callback should be part of the content. Currently nothing happens :(

My2cents :)

**Comment by kanariezwart** [comment]
@ro0NL I don't like it either, but re-decoding is the quickest solution with existing code.
I think the JsonResponse object should just be a container for data. And a ```ResponseBuilder``` object should responsible for adding/modifying data and returning it when called by a viewHandler

**Comment by kanariezwart** [comment]
Please look at https://github.com/symfony/symfony/pull/16551

**Comment by ro0NL** [comment]
How about a `JsonResponse::createFromString` which parses callback from contents... `getData` decodes the content string lazy.

`JsonResponse::create` could also be renamed to createFromData so its signature doesnt need to be changed..

**Comment by javiereguiluz** [comment]
Closing it as fixed by #19552 where we added `JsonResponse::fromJsonString()`
